packages/edge-gateway/src/gateway.js

```diff
@@ -525,3 +525,20 @@ function getDurableRequestUrl(request, route, data) {
     body: data && JSON.stringify(data),
   })
 }
+
+/**
+ * Transforms race response with custom headers.
+ * Content-Security-Policy header specified to only allow requests within same origin.
+ *
+ * @param {Response} response
+ */
+function getTransformedResponseWithCustomHeaders(response) {
+  const clonedResponse = new Response(response.body, response)
+
+  clonedResponse.headers.set(
+    'content-security-policy',
+    "connect-src 'self'; script-src 'self'"
```

Comment:
I think we should use [`default-src`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) otherwise embedded content could be used to sidestep restrictions here e.g. embed image from URL passing secret as query param or part of the path, font, css etc.... 

```suggestion
    "default-src 'self'; script-src 'self'"
```

1. We should utilize [report-uri](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/report-uri) directive to collect data about what resources get blocked & inform our future decisions by it.
1. I don't know how common it is, but I think there is no reason to not allow `unsafe-inline` (along with `self`) so that inline scripts, css will continue to work. Without it script tags and style tags will be ignored unless I'm mistaken. 
2. We may also want to enable `unsafe-eval`,  I don't see a reason to block JS eval as some pages may be using them. On the other hand we can also not enable and see if they get reported.
3. I also think `data:*` and `blob:*` URLs could be enabled as those can not be utilized for data smuggling.
4. I think `default-src` does not cover [`navigato-to`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/navigate-to) so we should restrict that as well, otherwise attacker can simply make a submit button be a link that takes you to adversaries site delivering stolen secret as query param or even a URL path fragment.
5. It looks like `default-src` does not cover [`form-action`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy/form-action) either which again could be used to steal data via HTML forms.

I think if we add all the above we should cover all the possibilities, but then again CSPs are notoriously complicated so I think we should have at some tests to verify desired behavior.

Reply:
Thanks for all these details @Gozala ❤️ 

1. Definitely, I will create an issue to get that part of [Malware mitigation initiative](https://www.notion.so/SI1-Malware-Mitigation-510646f8057743ec80b6e7c6fa3edd0d) given we need to add an endpoint, persist data and see how to act from there
2. Agreed, adding that ✅ 
3. it should be fine to also add, but yeah let's wait and see if issues are reported
4. makes total sense thanks, didn't find these in mdn docs but now I see they are part of the spec ✅ 
5. yeah, `default-src` follows only `*-src`. `navigate-to` looks experimental and [not supported](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy#browser_compatibility) in any browser yet. I will add anyway as is part of the spec ✅
6.  ✅

> I think if we add all the above we should cover all the possibilities, but then again CSPs are notoriously complicated so I think we should have at some tests to verify desired behavior.

Definitely, I want to try a few of the reported CIDs with this and have a look on the behaviour. Once we have it in staging, I will be able to test it. Later on, we can actually consider having real tests part of the test suite, but this will be quite tricky and time consuming for now

Reply:
Added work track for `report-uri` at 2nd point in https://hackmd.io/@vasco-santos/rJWZsz-Cc